localstack/services/s3/presigned_url.py

```diff
@@ -206,6 +217,55 @@ def _get_op_from_request(self, request: Request):
             return self._service.operation_model("GetObject")
 
 
+def get_credentials_from_parameters(parameters: dict) -> PreSignedCredentials:
+    """
+    Extract and retrieves the credentials from the passed signed requests parameters (can be from the query string or
+    the form for POST requests)
+    :param parameters:
+    :return:
+    """
+    # This is V2 signature AccessKeyId
+    if not (access_key_id := parameters.get("AWSAccessKeyId")):
+        # If not present, then it is a V4 signature (casing differs between QS parameters and form)
+        credential_value = parameters.get(
+            "X-Amz-Credential", parameters.get("x-amz-credential", "")
+        ).split("/")
+        if len(credential_value):
+            access_key_id = credential_value[0]
+
+    if not access_key_id:
+        # fallback to the hardcoded value
+        access_key_id = DEFAULT_PRE_SIGNED_ACCESS_KEY_ID
+
+    if not (secret_access_key := get_secret_access_key_from_access_key_id(access_key_id)):
+        # if we could not retrieve the secret access key, it means the access key was not registered in LocalStack,
+        # fallback to hardcoded necessary secret access key
+        secret_access_key = DEFAULT_PRE_SIGNED_SECRET_ACCESS_KEY
+
+    security_token = parameters.get("X-Amz-Security-Token", None)
+    return PreSignedCredentials(access_key_id, secret_access_key, security_token)
+
+
+@cache
+def get_secret_access_key_from_access_key_id(access_key_id: str) -> Optional[str]:
+    """
+    We need to retrieve the internal secret access key in order to also sign the request on our side to validate it
+    For now, we need to access Moto internals, as they are no public APIs to retrieve it for obvious reasons.
+    If the AccessKey is not registered, use the default `test` value that was historically used for pre-signed URLs, in
+    order to support default use cases
+    :param access_key_id: the provided AccessKeyID in the Credentials parameter
+    :return: the linked secret_access_key to the access_key
+    """
+    from moto.iam.models import AccessKey, iam_backends
+
+    account_id = get_account_id_from_access_key_id(access_key_id)
```

Comment:
Q: Is it possible to use `STS:GetSessionToken` at this point and obtain the access key ID + secret access key programatically?

Reply:
I don't think so, I think `GetSessionToken` returns a set of temporary credentials which will be then be different than what was used to signed the request on the client side.

I think you would also need to create the client for STS with the Access Key ID and the Secret Access Key of the client?

Reply:
This made me think of @dfangl hackathon, where he must had to implement a similar logic to validate the signature of every request:
https://github.com/localstack/localstack/commit/f7583642a0c78215dd742880f290411db93836f6

https://github.com/localstack/localstack/compare/master...hackathon-sig

And I guess it was the same issue, about having credentials that are set in LocalStack.

So I'm just not sure about the reverse of the access key id <-> account id to access the store, or I need to iterate in the same way.

Reply:
There seems to be no API for getting the secret key, so I think this is the best way for now :)